File: ios/Sources/BackgroundLocationPlugin/KeyValueStore.swift
import Foundation

/// Minimal key-value persistence abstraction.
///
/// `TrackingStateStore` depends on this protocol instead of `UserDefaults` directly so
/// its logic can be unit-tested with an in-memory implementation (Dependency
/// Inversion), matching the Android plugin's `KeyValueStore` interface.
protocol KeyValueStore {
    func string(forKey key: String) -> String?
    func set(_ value: String?, forKey key: String)

    func bool(forKey key: String) -> Bool
    func set(_ value: Bool, forKey key: String)

    func double(forKey key: String) -> Double
    func set(_ value: Double, forKey key: String)

    func removeObject(forKey key: String)
}

/// `KeyValueStore` backed by `UserDefaults` — the production implementation.
final class UserDefaultsKeyValueStore: KeyValueStore {
    private let defaults: UserDefaults

    init(suiteName: String = "com.elsapiens.backgroundlocation.state") {
        self.defaults = UserDefaults(suiteName: suiteName) ?? .standard
    }

    func string(forKey key: String) -> String? {
        defaults.string(forKey: key)
    }

    func set(_ value: String?, forKey key: String) {
        defaults.set(value, forKey: key)
    }

    func bool(forKey key: String) -> Bool {
        defaults.bool(forKey: key)
    }

    func set(_ value: Bool, forKey key: String) {
        defaults.set(value, forKey: key)
    }

    func double(forKey key: String) -> Double {
        defaults.double(forKey: key)
    }

    func set(_ value: Double, forKey key: String) {
        defaults.set(value, forKey: key)
    }

    func removeObject(forKey key: String) {
        defaults.removeObject(forKey: key)
    }
}

File: android/src/main/java/com/elsapiens/backgroundlocation/KeyValueStore.java
package com.elsapiens.backgroundlocation;

/**
 * Minimal key-value persistence abstraction.
 *
 * {@link TrackingStateStore} depends on this interface instead of Android's
 * SharedPreferences directly so its logic can be unit-tested with an in-memory
 * implementation (Dependency Inversion).
 */
public interface KeyValueStore {
    String getString(String key, String defaultValue);

    void putString(String key, String value);

    boolean getBoolean(String key, boolean defaultValue);

    void putBoolean(String key, boolean value);

    long getLong(String key, long defaultValue);

    void putLong(String key, long value);

    float getFloat(String key, float defaultValue);

    void putFloat(String key, float value);

    void remove(String key);
}
